### DISCUSSION TITLE
Does KeyCloak support scopes in docker-v2 client?

### DISCUSSION THREAD

**Discussion by FlyveHest** [keycloak/keycloak#12027] [Q&A]
Its pretty hard to read from the [documentation](https://www.keycloak.org/docs/latest/server_admin/#_docker)

> It deviates by using a very specific JSON format for requests and responses as well as mapping repository names and permissions to the OAuth scope mechanism.

But, I can't find anything regarding setting up scope mappings so I can limit access to the registry, there is only one mapper, **Docker Auth Mapper** with a setting of **Allow All**, and no obvious way of limiting anything.

Has anyone gotten this to work, or can KeyCloak only be used for authentication against a docker registry?

**Comment by krombel** [comment]
Yes, this is true.

Some research brought me [to the implementation of the docker protocol](https://github.com/josh-cain/keycloak/tree/master/services/src/main/java/org/keycloak/protocol/docker). It seems [there](https://github.com/keycloak/keycloak/tree/main/services/src/main/java/org/keycloak/protocol/docker/mapper) is only the "[Allow All](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/docker/mapper/AllowAllDockerProtocolMapper.java)" mapper implemented.

=> You are right: For now you can only use KeyCloak for authentication but not for implementing scope based filtering (as documented [here](https://docs.docker.com/registry/spec/auth/token/#requesting-a-token)).

**Reply by alexanderwolz** [reply]
Hi @krombel 

I exactly had the same issue with the docker registry authentication. So I ended up writing my own mapper. I use Keycloak 21 and registry 2.8.2. My mapper is configurative by environment variables and you can group users according to their email domain, roles and groups.
Check out https://github.com/alexanderwolz/keycloak-docker-group-role-mapper if it suits your needs.

**Reply by FlyveHest** [reply]
@alexanderwolz Looks very promising, i'll take a look and see if it solves my issues.
